### PR TITLE
Create indexes on Continuous Aggregates

### PR DESCRIPTION
**Pull request by fabriziomello**
Timescale 2.7 released a new version of Continuous Aggregate (#4269)
that allows users efectivelly create and use indexes in the
materialization hypertable. The boring part of it is that users should
discover what is the associated materialization hypertable to issue a
`CREATE INDEX` statement.
    
Improved it by allowing users to easily create indexes in the
materialization hypertable by simple executing a `CREATE INDEX` direct
in the Continuous Aggregate.
    
Example:
  `CREATE INDEX name_of_the_index ON continuous_agregate (column);`

### CODE REVIEW

**Review by svenklemm — COMMENTED** [review]
Creating unique indexes should probably blocked as you can mess up your cagg that way, also needs tests with unique indexes

**Review by mkindahl — COMMENTED** [review]
I think you're missing tests for the following cases:
- Using `ONLY`. It should default to the same situation as for hypertables, but better check.
- Using `UNIQUE`. Adding `UNIQUE` can cause the refresh job to start failing, so not sure how useful it is, but not convinced that is a good reason to now allow it.
- Using `CONCURRENTLY`. We don't support it on hypertables right now, but better check that it gives an error.
- Test for partial indexes.
- Using `INCLUDE`. This can be quite useful, so good to add a check that it actually works so that we do not make a mistake causing it to not work in the future.

File: src/process_utility.c [hunk 1]
```diff
@@ -2568,8 +2568,34 @@ process_index_start(ProcessUtilityArgs *args)
...
+		if (NULL == ht)
+		{
+			ts_cache_release(hcache);
+			return DDL_CONTINUE;
+		}
+
+		/* Make the RangeVar for the underlying materialization hypertable */
+		stmt->relation = makeRangeVar((char *) NameStr(ht->fd.schema_name),
+									  (char *) NameStr(ht->fd.table_name),
+									  0);
```
Comment (mfundul) [inline src/process_utility.c:2598]:
```suggestion
									  -1);
```
Shouldn't token location be unknown?

File: src/process_utility.c [hunk 2]
```diff
@@ -2568,8 +2568,34 @@ process_index_start(ProcessUtilityArgs *args)
...
+
+		if (NULL == ht)
+		{
+			ts_cache_release(hcache);
+			return DDL_CONTINUE;
+		}
+
+		/* Make the RangeVar for the underlying materialization hypertable */
+		stmt->relation = makeRangeVar((char *) NameStr(ht->fd.schema_name),
+									  (char *) NameStr(ht->fd.table_name),
```
Comment (mfundul) [inline src/process_utility.c:2597]:
```suggestion
		stmt->relation = makeRangeVar(NameStr(ht->fd.schema_name),
									  NameStr(ht->fd.table_name),
```

File: src/process_utility.c [hunk 3]
```diff
@@ -2568,8 +2568,34 @@ process_index_start(ProcessUtilityArgs *args)
...
+			}
+		}
+
+		if (NULL == ht)
+		{
+			ts_cache_release(hcache);
+			return DDL_CONTINUE;
+		}
+
+		/* Make the RangeVar for the underlying materialization hypertable */
```
Comment (mfundul) [inline src/process_utility.c:2595]: In general, I think this `stmt->relation = makeRangeVar` should only happen for the new case of CAGGs, and not for all hypertables. It could be harmless, but still, it looks like an unnecessary change for the common codepath to me.

File: src/process_utility.c [hunk 4]
```diff
@@ -2568,8 +2568,34 @@ process_index_start(ProcessUtilityArgs *args)
...
+			/* If the relation is a CAgg and it is finalized */
+			if (ContinuousAggIsFinalized(cagg))
+				ht = ts_hypertable_get_by_id(cagg->data.mat_hypertable_id);
+			else
+			{
+				ts_cache_release(hcache);
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("operation not supported on continuous aggreates that are not "
+								"finalized")));
```
Comment (mfundul) [inline src/process_utility.c:2585]: Maybe add a hint that the user needs to recreate the CAGG, since they may not understand what finalized is.

**Comment by fabriziomello** [comment]
> Creating unique indexes should probably blocked as you can mess up your cagg that way, also needs tests with unique indexes

Fixed!

**Comment by fabriziomello** [comment]
> I think you're missing tests for the following cases:
> 
> * Using `ONLY`. It should default to the same situation as for hypertables, but better check.
> * Using `UNIQUE`. Adding `UNIQUE` can cause the refresh job to start failing, so not sure how useful it is, but not convinced that is a good reason to now allow it.
> * Using `CONCURRENTLY`. We don't support it on hypertables right now, but better check that it gives an error.
> * Test for partial indexes.
> * Using `INCLUDE`. This can be quite useful, so good to add a check that it actually works so that we do not make a mistake causing it to not work in the future.

All fixed!

File: src/process_utility.c (hunk 1 above)
Comment (fabriziomello) [reply src/process_utility.c:2598]: Fixed!

File: src/process_utility.c (hunk 2 above)
Comment (fabriziomello) [reply src/process_utility.c:2597]: Fixed!

File: src/process_utility.c (hunk 4 above)
Comment (fabriziomello) [reply src/process_utility.c:2585]: Added a hint!

File: src/process_utility.c (hunk 3 above)
Comment (fabriziomello) [reply src/process_utility.c:2595]: But it will happen only for the new CAGG. I've added a test before and raise an error if the CAGG is not finalized!

Comment (mfundul) [reply src/process_utility.c:2601]: Yes, I've misread the code, my bad.

**Review by mkindahl — APPROVED** [review]

**Review by svenklemm — APPROVED** [review]
